step-67: explicit time integration with matrix-free implementations

kronbichler opened the pull request:
This tutorial implements an explicit time integrator with matrix-free implementation for DG elements. A Runge-Kutta scheme is used. A particular feature presented here is the [MatrixFreeOperators::CellwiseInverseMassMatrix class](https://www.dealii.org/developer/doxygen/deal.II/classMatrixFreeOperators_1_1CellwiseInverseMassMatrix.html) and a few other things that are not included in the Poisson-type problems we have for demonstrating matrix-free algorithms right now.

As equations, I implemented the Euler equations with some setup that does not lead to too many shocks in order to avoid implementing shock capturing, which should end up being maybe ~600-1000 lines of code (including comments)~[update: I wrote lots of comments, so we are now at around 1k lines of code only and 2300 lines of code including comments].

~Scalar advection would also be possible but is a bit basic, given that we have several simple Poisson operators for the matrix-free case already. Acoustic wave propagation would have been another possibility but I have another project [ExWave](https://github.com/kronbichler/exwave) already that explains this, so it feels redundant. Vote with :+1: if you agree on the Euler example or :confused: if you want something else.~

tjhei commented:
> @tjhei I guess you had a very different setup in mind for the nonlinear matrix-free step-66 in #8229?

yes, I don't think there would be any overlap. I agree that the simplest, interesting PDE makes sense, so your plan sounds solid! :+1:

bangerth commented:
I suspect you want this merged to reserve the number? If so, please merge yourself.

kronbichler commented:
> I suspect you want this merged to reserve the number? If so, please merge yourself.

I wanted to collect some opinions and communicate it early. I also do actually plan to write this before the summer break, which would be the beginning of August for me. I have not inserted the unfinished tag yet, so I will not merge at this point in the hope to finish it within reasonable time.

kronbichler commented:
/rebuild

kronbichler commented:
This tutorial is now ready for review. I still need to create some nice pictures of the sound waves of the channel flow around a cylinder test case, add some compute times for that setup, and write a few things one could do as extension, but apart everything is ready and I would fill in the missing data once we have agreed on the overall direction of the tutorial.

peterrum commented:
I would like to review this one. I have downloaded the branch und built the doxygen documentation. Unfortunately the latex equations are not rendered and are shown as pure code. Any suggestions to make it work?

peterrum commented on examples/step-67/step-67.cc:
```diff
@@ -0,0 +1,2305 @@
+
+
+
+  // This function implements the inverse mass matr
```
This is not updated yet, is it?

kronbichler replied:
I'm waiting for PR #9231 to be reviewed and merged, then I will add that code.

peterrum commented on examples/step-67/step-67.cc:
```diff
@@ -0,0 +1,2305 @@
+
+
+
+  // This function implements the inverse mass matr
```
This would generally a useful function in the library.

peterrum replied:
The same here.

kronbichler commented on examples/step-67/step-67.cc:
```diff
@@ -0,0 +1,2305 @@
+  // variables, using the formula $p = (\gamma - 1) \left(E - \frac 12 \rho
+  // \mathbf{u}\cdot \mathbf{u}\right)$. As explained above, we use the
+  // velocity from the `euler_velocity()` function. Note that we need to
+  // specify the first template argument `dim` here because the compiler is
```
No, the compiler can't deduced the template parameter from a derived value, see e.g. https://godbolt.org/z/pb7YDL
(you can comment in the offending line without template). It works if you would have the argument `Tensor<1, dim, Number>`, but I can't do that here. Well, I could make the argument "dim_plus_two" and deduce the rest, but that is worse.

peterrum replied:
This surprises me! Thanks! I guess how it is, its better.

kronbichler commented on examples/step-67/step-67.cc:
```diff
@@ -0,0 +1,2305 @@
+    const auto pressure_p = euler_pressure<dim>(u_p);
+
+    const auto flux_m = euler_flux<dim>(u_m);
+    const auto flux_p = euler_flux<dim>(u_p);
```
It does indeed, at least gcc. Here is the assembly output of the loop over quadrature points for `local_apply_face`, which includes the numerical flux:
```assembly
.L11816:
	movq	-25560(%rbp), %r15
	vmovapd	(%r15), %zmm6 ; normal vector x
	vmovapd	64(%r15), %zmm5 ; normal vector y
	vmovapd	128(%r15), %zmm4 ; normal vector z
.L11772:
	movq	-25464(%rbp), %rsi
	movq	-25496(%rbp), %r12
	movq	-25520(%rbp), %r15
	leaq	0(,%rdx,8), %rax
	movq	-25472(%rbp), %r14
	movq	-25504(%rbp), %r11
	leaq	(%rsi,%rax), %r10
	leaq	(%r12,%rax), %rsi
	leaq	(%r15,%rax), %r12
	vmovapd	(%r12), %zmm13
	vmovapd	.LC70(%rip), %zmm0
	movq	-25480(%rbp), %r13
	movq	-25512(%rbp), %rbx
	leaq	(%r14,%rax), %r9
	leaq	(%r11,%rax), %r14
	vdivpd	(%r14), %zmm0, %zmm22 ; inverse density interior side
	vmulpd	%zmm13, %zmm22, %zmm16
	leaq	0(%r13,%rax), %r8
	leaq	(%rbx,%rax), %r13
	vmovapd	0(%r13), %zmm1
	movq	-25528(%rbp), %rbx
	vmulpd	%zmm1, %zmm22, %zmm10
	vmulpd	%zmm13, %zmm16, %zmm23
	addq	%rax, %rbx
	vmovapd	(%rbx), %zmm12
	movq	-25536(%rbp), %r11
	vmovapd	%zmm10, %zmm17
	vmulpd	%zmm12, %zmm22, %zmm2 ; velocity interior side ready
	vfmadd132pd	%zmm1, %zmm23, %zmm17
	vmovapd	%zmm12, %zmm9
	addq	%rax, %r11
	vmovapd	(%r11), %zmm26
	vmovapd	(%r8), %zmm14
	vfmadd132pd	%zmm2, %zmm17, %zmm9
	vfmadd231pd	%zmm12, %zmm2, %zmm17
	vmulpd	%zmm1, %zmm10, %zmm20 
	vdivpd	(%r10), %zmm0, %zmm21 ; inverse density exterior side
	vmulpd	%zmm14, %zmm21, %zmm15
	vmovapd	(%r9), %zmm3
	vfnmadd132pd	%zmm25, %zmm26, %zmm17
	vmulpd	%zmm3, %zmm21, %zmm8
	movq	-25488(%rbp), %rdi
	vmulpd	%zmm14, %zmm15, %zmm19
	addq	%rax, %rdi
	vfmadd231pd	%zmm24, %zmm17, %zmm20
	vmovapd	(%rdi), %zmm18
	vmulpd	%zmm10, %zmm12, %zmm30
	vfmadd231pd	%zmm24, %zmm17, %zmm23
	vmovapd	%zmm8, %zmm11
	vmovapd	%zmm20, -25008(%rbp)
	vmovapd	%zmm17, %zmm20
	vfmadd132pd	%zmm24, %zmm26, %zmm20
	vmulpd	%zmm18, %zmm21, %zmm7 ; velocity exterior side ready
        ; what comes next should be the evaluation of flux_m and flux_p (already started before)
	vfmadd132pd	%zmm3, %zmm19, %zmm11
	vmovapd	%zmm30, -24944(%rbp)
	vmovapd	%zmm23, %zmm30
	vmulpd	%zmm20, %zmm16, %zmm23
	vmulpd	%zmm20, %zmm10, %zmm28
	vfmadd231pd	%zmm18, %zmm7, %zmm11
	vmulpd	%zmm16, %zmm12, %zmm29
	vmulpd	%zmm13, %zmm10, %zmm31
	vmovapd	%zmm23, -25200(%rbp)
	vmulpd	%zmm2, %zmm12, %zmm23
	vfnmadd213pd	(%rsi), %zmm25, %zmm11
	vmovapd	%zmm28, -25072(%rbp)
	vmovapd	%zmm29, -25136(%rbp)
	vmulpd	%zmm1, %zmm2, %zmm28
	vfmadd132pd	%zmm24, %zmm23, %zmm17
	vmulpd	%zmm24, %zmm11, %zmm0
	vmulpd	%zmm13, %zmm2, %zmm29
	vfnmadd132pd	%zmm25, %zmm26, %zmm9
	vfmadd132pd	%zmm10, %zmm27, %zmm10
	vmovapd	%zmm17, -25392(%rbp)
	vmulpd	%zmm20, %zmm2, %zmm17
	vmulpd	%zmm14, %zmm8, %zmm20
	vmovapd	%zmm0, -24496(%rbp)
	vmulpd	%zmm3, %zmm7, %zmm0
	vmovapd	%zmm31, -24880(%rbp)
	vmovapd	%zmm28, -25264(%rbp)
	vmovapd	%zmm29, -25328(%rbp)
	vmovapd	%zmm17, -25456(%rbp)
	vmovapd	%zmm20, -24560(%rbp)
	vmovapd	%zmm11, %zmm20
	vfmadd213pd	(%rsi), %zmm24, %zmm20
	vmovapd	%zmm0, -24752(%rbp)
	vmulpd	%zmm24, %zmm9, %zmm9
	vmulpd	%zmm14, %zmm7, %zmm0
	vfmadd231pd	%zmm16, %zmm16, %zmm10
	vmulpd	%zmm18, %zmm8, %zmm28
	vmulpd	%zmm20, %zmm8, %zmm29
	vmulpd	%zmm3, %zmm8, %zmm17
	vmulpd	.LC82(%rip), %zmm9, %zmm9
	vmovapd	%zmm0, -24816(%rbp)
	vfmadd132pd	%zmm8, %zmm27, %zmm8
	vmulpd	%zmm18, %zmm7, %zmm0
	vfmadd231pd	%zmm2, %zmm2, %zmm10
	vmovapd	-24496(%rbp), %zmm2
	vfmadd231pd	%zmm24, %zmm11, %zmm17
	vfmadd231pd	%zmm24, %zmm11, %zmm19
	vfmadd231pd	%zmm15, %zmm15, %zmm8
	vfmadd132pd	%zmm24, %zmm0, %zmm11
	vmulpd	%zmm22, %zmm9, %zmm0 ; use inverse density m for lambda
	vmulpd	.LC82(%rip), %zmm2, %zmm9
	vsubpd	%zmm14, %zmm13, %zmm2
	vfmadd231pd	%zmm7, %zmm7, %zmm8
	vmulpd	%zmm14, %zmm5, %zmm14
	vaddpd	%zmm10, %zmm0, %zmm10
	vmulpd	%zmm21, %zmm9, %zmm0  ; use inverse density p for lambda
	vmulpd	%zmm1, %zmm16, %zmm31
	vmulpd	%zmm20, %zmm15, %zmm23
	vfmadd231pd	%zmm3, %zmm6, %zmm14
	vmovapd	%zmm28, -24624(%rbp)
	vaddpd	%zmm8, %zmm0, %zmm8
	vmulpd	%zmm18, %zmm15, %zmm28
	vmovapd	%zmm29, -24688(%rbp)
	vmulpd	%zmm3, %zmm15, %zmm29
	vmaxpd	%zmm10, %zmm8, %zmm15 ; max for lambda
	vmovapd	(%r14), %zmm10
	vfmadd231pd	%zmm18, %zmm4, %zmm14
	vmulpd	%zmm31, %zmm5, %zmm21
	vsubpd	%zmm18, %zmm12, %zmm0
	vmulpd	%zmm20, %zmm7, %zmm20
	vmulpd	%zmm29, %zmm5, %zmm18
	vsqrtpd	%zmm15, %zmm7 ; sqrt for lambda
        ; the following operations involving zmm4, zmm5, zmm6 multiply the fluxes by the normal
	vmulpd	%zmm28, %zmm5, %zmm22
	vsubpd	%zmm3, %zmm1, %zmm15
	vmulpd	%zmm23, %zmm5, %zmm16
	vmulpd	%zmm13, %zmm5, %zmm3
	vmulpd	-25136(%rbp), %zmm5, %zmm31
	vsubpd	(%r10), %zmm10, %zmm8
	vsubpd	(%rsi), %zmm26, %zmm10
	vmulpd	%zmm19, %zmm5, %zmm26
	vmulpd	%zmm30, %zmm5, %zmm19
	vmulpd	-25200(%rbp), %zmm5, %zmm5
	vmulpd	%zmm25, %zmm7, %zmm9
	vfmadd231pd	%zmm17, %zmm6, %zmm18
	vfmadd132pd	%zmm6, %zmm3, %zmm1
	vfmadd231pd	-24560(%rbp), %zmm6, %zmm26
	vfmadd231pd	-24624(%rbp), %zmm6, %zmm22
	vfmadd231pd	-24688(%rbp), %zmm6, %zmm16
	vfmadd231pd	-25008(%rbp), %zmm6, %zmm21
	vfmadd231pd	-24880(%rbp), %zmm6, %zmm19
	vfmadd231pd	-24944(%rbp), %zmm6, %zmm31
	vfmadd132pd	-25072(%rbp), %zmm5, %zmm6
	vmulpd	%zmm25, %zmm9, %zmm9
	vfmadd132pd	%zmm4, %zmm1, %zmm12
	vfmadd132pd	%zmm4, %zmm22, %zmm11
	vfmadd231pd	-24752(%rbp), %zmm4, %zmm18
	vfmadd231pd	-24816(%rbp), %zmm4, %zmm26
	vfmadd132pd	%zmm4, %zmm16, %zmm20
	vfmadd231pd	-25264(%rbp), %zmm4, %zmm21
	vfmadd231pd	-25328(%rbp), %zmm4, %zmm19
	vfmadd231pd	-25392(%rbp), %zmm4, %zmm31
	vfmadd132pd	-25456(%rbp), %zmm6, %zmm4
        ; compute lambda * (u_m - u_p)
	vmulpd	%zmm15, %zmm9, %zmm7
	vmulpd	%zmm8, %zmm9, %zmm8
	vmulpd	%zmm2, %zmm9, %zmm15
	vmulpd	%zmm0, %zmm9, %zmm0
	vmulpd	%zmm10, %zmm9, %zmm9
        ; add flux_m * normal + flux_p * normal
	vaddpd	%zmm14, %zmm12, %zmm6
	vaddpd	%zmm11, %zmm31, %zmm1
	vaddpd	%zmm20, %zmm4, %zmm29
	vaddpd	%zmm18, %zmm21, %zmm30
	vaddpd	%zmm26, %zmm19, %zmm17
        ; add 0.5 * [lambda * (u_m - u_p)] to 0.5 * (flux_m * normal + flux_p * normal)
	vfmadd231pd	%zmm6, %zmm25, %zmm8
	vfmadd231pd	%zmm30, %zmm25, %zmm7
	vfmadd231pd	%zmm17, %zmm25, %zmm15
	vfmadd231pd	%zmm1, %zmm25, %zmm0
	vfmadd231pd	%zmm29, %zmm25, %zmm9
	cmpb	$1, -24420(%rbp)
        ; switch sign for flux on m side
	vsubpd	%zmm8, %zmm27, %zmm13
	vsubpd	%zmm7, %zmm27, %zmm12
	vsubpd	%zmm15, %zmm27, %zmm4
	vsubpd	%zmm0, %zmm27, %zmm2
	vsubpd	%zmm9, %zmm27, %zmm11
        ; the following jumps are for the affine mesh/general mesh case on the two sides
	jbe	.L11814
        ; this is the general mesh case (different JxW in each point) for eval_m
	movq	-25552(%rbp), %r15
	vmovapd	(%r15,%rdx,8), %zmm10
	vmulpd	%zmm13, %zmm10, %zmm14
	vmulpd	%zmm12, %zmm10, %zmm3
	vmulpd	%zmm4, %zmm10, %zmm5
	vmulpd	%zmm2, %zmm10, %zmm6
	vmulpd	%zmm11, %zmm10, %zmm1
	vmovapd	%zmm14, (%r14)
	vmovapd	%zmm3, 0(%r13)
	vmovapd	%zmm5, (%r12)
	vmovapd	%zmm6, (%rbx)
	vmovapd	%zmm1, (%r11)
.L11774:
	cmpb	$1, -23172(%rbp)
	jbe	.L11815
        ; this is the general mesh case (different JxW in each point) for eval_m
	movq	-25544(%rbp), %r14
	vmovapd	(%r14,%rax), %zmm28
	vmulpd	%zmm8, %zmm28, %zmm8
	vmulpd	%zmm7, %zmm28, %zmm7
	vmulpd	%zmm15, %zmm28, %zmm15
	vmulpd	%zmm0, %zmm28, %zmm0
	vmulpd	%zmm9, %zmm28, %zmm9
	vmovapd	%zmm8, (%r10)
	vmovapd	%zmm7, (%r9)
	vmovapd	%zmm15, (%r8)
	vmovapd	%zmm0, (%rdi)
	vmovapd	%zmm9, (%rsi)
.L11776:
	addq	$8, %rdx
	addq	$192, %rcx
	cmpq	%rdx, -24432(%rbp)
	je	.L11777
.L11778:
	cmpb	$2, -24420(%rbp)
	jbe	.L11816
	vmovapd	(%rcx), %zmm6 ; normal vector x for next round
	vmovapd	64(%rcx), %zmm5 ; normal vector y for next round
	vmovapd	128(%rcx), %zmm4 ; normal vector y for next round
	jmp	.L11772
	.p2align 4,,10
	.p2align 3
.L11815:
        ; this multiplies by JxW for the Cartesian/affine mesh case
	movq	-23200(%rbp), %rax
	movq	-25544(%rbp), %r13
	vbroadcastsd	(%rax,%rdx), %zmm3
	vmulpd	0(%r13), %zmm3, %zmm5
	vmulpd	%zmm8, %zmm5, %zmm6
	vmulpd	%zmm7, %zmm5, %zmm1
	vmulpd	%zmm15, %zmm5, %zmm10
	vmulpd	%zmm0, %zmm5, %zmm14
	vmulpd	%zmm9, %zmm5, %zmm13
	vmovapd	%zmm6, (%r10)
	vmovapd	%zmm1, (%r9)
	vmovapd	%zmm10, (%r8)
	vmovapd	%zmm14, (%rdi)
	vmovapd	%zmm13, (%rsi)
	jmp	.L11776
	.p2align 4,,10
	.p2align 3
.L11814:
        ; this multiplies by JxW for the Cartesian/affine mesh case
	movq	-23856(%rbp), %r15
	vbroadcastsd	(%r15,%rdx), %zmm10
	movq	-25552(%rbp), %r15
	vmulpd	(%r15), %zmm10, %zmm14
	vmulpd	%zmm13, %zmm14, %zmm13
	vmulpd	%zmm12, %zmm14, %zmm12
	vmulpd	%zmm4, %zmm14, %zmm4
	vmulpd	%zmm2, %zmm14, %zmm2
	vmulpd	%zmm11, %zmm14, %zmm11
	vmovapd	%zmm13, (%r14)
	vmovapd	%zmm12, 0(%r13)
	vmovapd	%zmm4, (%r12)
	vmovapd	%zmm2, (%rbx)
	vmovapd	%zmm11, (%r11)
	jmp	.L11774  
```
I have checked the formulas involved here (even though reading assembly is tough) and I cannot spot redundant operations here. There are exactly two divisions and one square root (for the modified Lax-Friedrichs flux), no function calls in between the quadrature point region, and jumps only due to the decision in `FEEvaluationBase::submit_value()` to distinguish the plus/minus side.

kronbichler replied:
Of course, this kind of assembly code is incredibly hard to read because the compiler schedules operations out of order. But overall, I think the assembly code is super clean.

peterrum replied:
Thanks: I am convinced!

kronbichler commented on examples/step-67/step-67.cc:
```diff
@@ -0,0 +1,2305 @@
+          result[d][v] = function.value(p, d);
+      }
+    return result;
+  }
```
Yes, we should. The question is how we want to define the multi-component case and in which file we want to put it. It looks like it should belong to `include/deal.II/base/function.h`, but I have not thought much about whether that would pull in `vectorization.h` in an even more central place than what we already do. Let us keep the code here for now and open an issue.

kronbichler replied:
The issue is #9269.

kronbichler commented on examples/step-67/step-67.cc:
```diff
@@ -0,0 +1,2305 @@
+    const Number           inverse_density = 1. / conserved_variables[0];
+    Tensor<1, dim, Number> velocity;
+    for (unsigned int d = 0; d < dim; ++d)
+      velocity[d] = conserved_variables[1 + d] * inverse_density;
```
You mean like we do in `step-33`? I can do that, but it makes code pretty long because I would write `inverse_density = Number(1.)/conserved_variables[density_component]` and then `velocity[d] = conserved_variables[first_velocity_component + d] * inverse_density;`. This will make the program 20-40 lines longer I suspect (as I'm close to the 80 column limit already now), but if there is consensus to do that, I will.

peterrum replied:
I think it would help to understand the code. You do not have to do for me: I am familiar with this ;)

kronbichler commented on examples/step-67/step-67.cc:
```diff
@@ -0,0 +1,2305 @@
+                     inflow_boundaries.end())
+              w_p =
+                evaluate_function(*inflow_boundaries.find(boundary_id)->second,
+                                  phi.quadrature_point(q));
```
Good point - indeed that's another of those places where I was speculating on the compiler to optimize the second call away. I don't like to read x86 integer assembly, so I have not checked whether it looks up the map twice or not. I thought a little bit about using some kind of `if (auto var = inflow_boundaries.find(boundary_id))` to assign the variable, but it did not work (as the map iterator is not convertible to bool; so I somehow don't see how to make the final comparison to `inflow_boundaries.end()`).

peterrum replied:
I tried the same ;) I am so annoyed by C++ at this point: more on this in another PR ;)

kronbichler commented on examples/step-67/step-67.cc:
```diff
@@ -0,0 +1,2305 @@
+            else
+              AssertThrow(false,
+                          ExcMessage("Unknown boundary id, did "
+                                     "you set a boundary condition?"));
```
I kind of have to, because I otherwise force the compiler to inline the numerical flux 4 times, and I would believe that duplicating the code is worse for the instruction cache. Thinking about it, I wonder I should remove the `always_inline` decoration from the numerical flux (but not the other places), so I could share the code among a few places. Well, I think it's better not to, as the compiler will likely mess things up.

kronbichler commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,686 @@
+effort, we refrain from the regime of the Euler equations with pronounced
+shocks, and rather concentrate on the regime of subsonic flows with wave-like
+phenomena. For a method that works well with shocks (but is more expensive per
+unknown), we refer to the step-69 tutorial program.
```
I guess it will be quite hard to match the per-dof performance you get with the plain DG code as soon as you start limiting, no matter what scheme you choose. Maybe artificial viscosity would be fastest in terms of DoF/s, but that is not nearly as good in terms of solution quality/crispness as the second-order variant of the approach from step-69.

kronbichler commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,686 @@
+(\rho \mathbf u)^- \\ E_\mathrm{D}(t)\end{pmatrix} \quad
+ \text{(mixed Neumann/Dirichlet)},
+@f]
+<b>supersonic outflow boundaries</b>, where all characteristics are outgoing
```
That's a question: I put it in because it is easy to implement, on the other hand all examples with supersonic flows I tried were having too many oscillations and failed due to negative density. If nobody objects, I will remove it and put the code to the possibilities for extensions.

peterrum replied:
This sounds like a good idea!

kronbichler commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,686 @@
+we use to evaluate the operator $\mathcal L_h$ and the inverse mass matrix
+$\mathcal M$. Actually, the term <i>matrix-free</i> is a slight misnomer,
+because we are working with a nonlinear operator and do not linearize the
+operator that in turn could be represented by a matrix. However, fast
```
You are correct, so let me simply add the comment that the inverse mass matrix is still matrix-free (whereas the residual is, well, a nonlinear operator evaluation).

kronbichler commented:
I have updated all code, except for the named positions for density, momentum and energy. I will likely add those as well, but wait a bit for other reviews. At that point, I will also be ready with the rest of the results.

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+eigenvalues of the linearization of $\mathbf F(\mathbf w)$ with respect to
+$\mathbf{w}$. In this program, we set the maximal time step as follows:
+@f[
+\Delta t = \frac{Cr}{p^{1.5}}\left(\frac{1}{\max\left[\frac{\|\mathbf{u}\|}{h_u} +
```
Do you use this as the *maximal* time step, or the *actual* time step size?

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+convective and acoustic limits, as the Euler equations can transport
+information with speed $\|\mathbf{u}\|+c$.
+
+In this tutorial program, we use a specific variant of explicit Runge--Kutta
```
Somewhere here it would probably be good to link to either https://en.wikipedia.org/wiki/Runge%E2%80%93Kutta_methods and/or the book by Hairer and Wanner (and Norsett in the latest edition).

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+@f]
+
+In this tutorial program, we use a subset of explicit Runge--Kutta methods,
+so-called low-storage Runge--Kutta methods (LSRK). The motivation for
```
out of curiosity, is storing a few extra vectors really a big deal? I mean, you're not going to use a 20-stage RK method :-)

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+\end{aligned}
+@f]
+Besides the vector $\mathbf w_h$ that is successively updated, this scheme
+only needs two auxiliary vectors, namely the vector $\mathbf{K}_i$ to hold the
```
By the way, if you were renaming cells to use the symbol `$K$`, you might want to also use `\mathbf{k}` with a lower-case `k` for the RK stages. That happens also to be the notation used at the wikipedia site.

kronbichler replied:
Changed to lower-case `\mathbf{k}`.

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+
+The main advantages of low-storage variants are the reduced memory consumption
+on the one hand and the reduced memory access on the other. In this program,
+we are particularly interested in the latter aspect. Since cost of operator
```
ah, I see :-)

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+
+Besides three variants for third, fourth and fifth order accuracy from the
+reference @cite KennedyCarpenterLewis2000, we also use a variant that was
+optimized for acoustics setups from @cite TseliosSimos2007. Acoustic problems
```
can you add which order (and how many stages) this method has?

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+the scalar systems in previous tutorial programs. In the code, all that
+changes is a template argument of the FEEvaluation and FEFaceEvaluation
+classes, the one to set the number of components. The access to the vector is
+the same as before, all handled transparently by the evalautor. We also note
```
```suggestion
the same as before, all handled transparently by the evaluator. We also note
```

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+the Laplacian of step-37, it now returns a type that is
+`Tensor<1,dim+2,VectorizedArray<Number>>`. Likewise, the gradient type is now
+`Tensor<1,dim+2,Tensor<1,dim,VectorizedArray<Number>>>`, where the outer
+tensor collects the `dim+2` components of the Euler system, and the the inner
```
```suggestion
tensor collects the `dim+2` components of the Euler system, and the inner
```

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+gradient of the test function, the integrand is of degree $3p$ in the
+numerator and $p$ in the denominator already for affine (undeformed) cell
+geometries. For curved cells, additional polynomial and rational expression
+appear. At this point, one usually needs to give up on insisting on exact
```
```suggestion
appear when multiplying the integrand by the determinant of the Jacobian of the 
mapping. At this point, one usually needs to give up on insisting on exact
```

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+geometries. For curved cells, additional polynomial and rational expression
+appear. At this point, one usually needs to give up on insisting on exact
+integration, and take whatever accuracy the Gaussian (more precisely,
+Gauss--Legrende) quadrature provides. As that formulas only integrate
```
I'll note that that is the same when you integrate for the Laplace equation: The term `grad phi_i . grad phi_j` on non-affine cells is a rational function because the gradient involves the inverse of the Jacobian matrix; this means dividing by the determinant of the Jacobian, which is a polynomial of degree `p*d`. In other words, for non-affine cells, we don't even integrate exactly for the Laplace equation, even though it is linear.

bangerth replied:
```suggestion
Gauss--Legrende) quadrature provides. As these formulas only integrate
```

kronbichler replied:
I added a short comment on the inexact integration of the Laplace equation on non-affine cells.

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+appear. At this point, one usually needs to give up on insisting on exact
+integration, and take whatever accuracy the Gaussian (more precisely,
+Gauss--Legrende) quadrature provides. As that formulas only integrate
+polynomials exactly, we have to live with the variational crime in the form of
```
maybe link to this for the "variational crime": https://mathoverflow.net/questions/26018/what-are-variational-crimes-and-who-coined-the-term

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+effect called <b>aliasing</b>. The term comes from signal processing and
+expresses the situation of inappropriate, too coarse sampling. In terms of
+quadrature, the inappropriate sampling means that we use too few quadrature
+points as what would be required to accurately sample the variable-coefficient
```
```suggestion
points compared to what would be required to accurately sample the variable-coefficient
```

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+expresses the situation of inappropriate, too coarse sampling. In terms of
+quadrature, the inappropriate sampling means that we use too few quadrature
+points as what would be required to accurately sample the variable-coefficient
+integrand. It has been shown in DG literature that aliasing errors can
```
```suggestion
integrand. It has been shown in the DG literature that aliasing errors can
```

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+points as what would be required to accurately sample the variable-coefficient
+integrand. It has been shown in DG literature that aliasing errors can
+introduce unphysical oscillations in the numerical solution for <i>barely</i>
+resolved simulations. The fact that aliasing mostly effects coarse resolutions
```
```suggestion
resolved simulations. The fact that aliasing mostly affects coarse resolutions
```

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+surprising because well-resolved simulations have small coefficients in the
+higher polynomial degrees that are missed by too few quadrature points,
+whereas the main solution contribution in the lower polynomial degrees is
+still well-captured. To address this topic, various approaches have been
```
In other words, Taylor's theorem helps you! On short enough length scales, every function looks linear and can easily be integrated accurately :-)

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+consistency), and transform back to the nodal basis. However, filters reduce
+the accuracy of the method. Another, in some sense simpler, strategy is to use
+more quadrature points to capture non-linear terms more accurately. Using more
+than $p+1$ quadrature points is sometimes called over-integration or
```
```suggestion
than $p+1$ quadrature points per coordinate direction is sometimes called over-integration or
```

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+exactly polynomials of degree $3p$ or $4p$, respectively) are common.
+
+To reflect this variability in the choice of quadrature in the program, we
+keep the number of quadrature a variable to be specified just as the
```
```suggestion
keep the number of quadrature points a variable to be specified just as the
```

bangerth commented on examples/step-67/doc/kind:
```diff
@@ -0,0 +1 @@
+techniques
```
maybe make it a fluid flow problem instead?

kronbichler replied:
Yes, fluid flow is good.

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+<h3>Evaluation of the inverse mass matrix with matrix-free techniques</h3>
+
+The last ingredient is the evaluation of the inverse mass matrix $\mathcal
+M^{-1}$. In DG methods with explicit time integration, mass matrices are
```
Did you make it mathcal above? I don't recall!

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+polynomials coincide with the points of quadrature, obviating some
+interpolation operations. Given the fact that we want to use more quadrature
+points for nonlinear terms in $\mathcal L_h$, however, the collocation
+property is lost. (More precisely, it is still used in FEEvaluation and
```
Well, one could use different quadrature formulas for different terms :-)

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+prefer a conventional Lagrange basis in Gauss-Lobatto points as those make the
+evaluation of face integrals cheap. One could of course also use the
+Gauss-Lobatto quadrature (with some additional integration error) as was done
+in step-48, but we do not want to sacrifice accuracy additionally at that
```
```suggestion
in step-48, but we do not want to sacrifice accuracy at that
```

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+which is more costly due to over-integration and face integrals. In fact, it
+is so cheap that it is limited by the bandwidth of reading the source vector,
+reading the diagonal, and writing into the destination vector on most modern
+architectures. The hardware used for in the result section allows to do the
```
```suggestion
architectures. The hardware used for the result section allows to do the
```

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+reading the diagonal, and writing into the destination vector on most modern
+architectures. The hardware used for in the result section allows to do the
+computations at least twice the throughput than what is possible due to
+streaming the vectors, given the bandwidth from RAM memory.
```
the grammar in the middle of the sentence is funny here...

bangerth commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+channel, using the GridGenerator::channel_with_cylinder() function. Here, we
+impose a subsonic initial field at Mach number of $\mathrm{Ma}=0.307$ with a
+constant velocity in $x$ direction. At the top and bottom wall as well as at
+the cylinder, we impose a no-penetration condition. This setup forces the flow
```
```suggestion
the cylinder, we impose a no-penetration (i.e., tangential flow) condition. This setup forces the flow
```

bangerth commented on examples/step-67/doc/results.dox:
```diff
@@ -0,0 +1,390 @@
+@code
+Running with 40 MPI processes
+Vectorization over 8 doubles = 512 bits (AVX512), VECTORIZATION_LEVEL=3
+Number of degrees of freedom: 589824 ( = 4 [vars] x 4096 [cells] x 36 [dofs/cell/var] )
```
If you use the approach here https://github.com/dealii/dealii/blob/master/examples/step-32/step-32.cc#L1873-L1879, then you get a comma separator:
```suggestion
Number of degrees of freedom: 589,824 ( = 4 [vars] x 4096 [cells] x 36 [dofs/cell/var] )
```

bangerth commented on examples/step-67/doc/results.dox:
```diff
@@ -0,0 +1,390 @@
+detects the acoustic limit in the later part of the simulation.
+
+The final summary shows that around 2500 time steps have been performed in 3.7
+seconds, while the output of 11 files has taken additional 3 seconds. Broken
```
:-)

bangerth commented on examples/step-67/doc/results.dox:
```diff
@@ -0,0 +1,390 @@
+down per time step, where five Runge--Kutta stages are performed, the compute
+time is 1.44 milliseconds. This high performance is typical of matrix-free
+methods and a reason why explicit time integration is very competitive against
+implicit solver especially for large-scale simulations. The breakdown of
```
```suggestion
implicit solvers, especially for large-scale simulations. The breakdown of
```

bangerth commented on examples/step-67/doc/results.dox:
```diff
@@ -0,0 +1,390 @@
+methods and a reason why explicit time integration is very competitive against
+implicit solver especially for large-scale simulations. The breakdown of
+computational times at the end of the output shows that the evaluation of
+$\mathcal L_h$ contribute with around 2.9 seconds, the application of the
```
```suggestion
$\mathcal L_h$ contributes with around 2.9 seconds, the application of the
```

bangerth commented on examples/step-67/doc/results.dox:
```diff
@@ -0,0 +1,390 @@
+   Spent 3.08s on output and 3.694s on computations.
+   Apply called 0 times
+   RK update called 12790 times
+   Evaluate L_h           :  2.763e+00 (p  39)  2.932e+00 3.137e+00 (p   0)
```
Can you add units to these numbers? Also, what is the 'p' value?

bangerth replied:
And what do the three columns indicate?

bangerth commented on examples/step-67/doc/results.dox:
```diff
@@ -0,0 +1,390 @@
+   Evaluate L_h           :  1.202e+02 (p   6)  1.257e+02 1.338e+02 (p  16)
+   Inverse mass (+RK upd) :  6.834e+01 (p  16)  7.884e+01 8.344e+01 (p   6)
+   Compute transport speed:  4.031e+00 (p  27)  6.658e+00 9.301e+00 (p  20)
+   Compute errors/norms:     4.206e-01 (p  27)  4.295e-01 4.390e-01 (p  31)
```
you'll want to align this last colon to the previous ones

bangerth commented on examples/step-67/doc/results.dox:
```diff
@@ -0,0 +1,390 @@
+for the Lax--Friedrichs flux for $p=2$, but the picture is reversed for
+$p=3$. For $p=5$, we reach the roundoff accuracy of $10^{-11}$ with both
+fluxes on the finest grids. Also note that the errors are absolute, and the
+domain length is $10^2$, so relative errors are below $10^{-12}$. The HLL flux
```
I assume that the solution is $O(1)$?

bangerth commented on examples/step-67/doc/results.dox:
```diff
@@ -0,0 +1,390 @@
+
+<h3>Results for flow in channel around cylinder in 2D</h3>
+
+<h3>Results for flow in channel around cylinder in 3D</h3>
```
I guess these are still missing?

bangerth commented on examples/step-67/doc/results.dox:
```diff
@@ -0,0 +1,390 @@
+                at_outflow = true;
+              }
+@endcode
+in the `local_apply_boundary_face()` function.
```
endline missing

bangerth commented on examples/step-67/doc/results.dox:
```diff
@@ -0,0 +1,390 @@
+In the code, we would simply add the additional statement
+@code
+            else if (supersonic_outflow_boundaries.find(boundary_id) !=
+                     supersonic_outflow_boundaries.end())
```
isn't it the case that whether or not an outflow boundary is supersonic depends on the solution, not just where you prescribe boundary indicators?

kronbichler commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+blocks of equations, the second involving $d$ components, describe the
+conservation of mass, momentum, and energy.  The system is closed by an
+equation for the pressure as $p=(\gamma - 1) \left(E-\frac 12 \rho
+\mathbf{u}\cdot \mathbf{u}\right)$ and the gas constant $\gamma = 1.4$.
```
I slightly rephrased it.

kronbichler commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+
+For the derivation of the DG formulation, we multiply the Euler equations with
+test functions $\mathbf{v}$ and integrate over an individual element
+$\Omega_e$, which gives
```
> In almost all of the tutorial programs, we use `K` for cells. Are you willing to reconsider for notational consistency?

Of course. I use `K` often myself, but somehow had the feeling that here I needed to switch... :-)

kronbichler commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+
+There is a large number of numerical flux functions available, also called
+Riemann solvers. For the Euler equations, there exist so-called exact Riemann
+solvers --- meaning that the states from both sides are combined in a way that
```
Let's go with two dashes then which we know works.

kronbichler commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+the local Lax--Friedrichs flux
+@f[
+\hat{\mathbf{F}}(\mathbf{w}^-,\mathbf{w}^+) =
+\frac{\mathbf{F}(\mathbf{w}^-)+\mathbf{F}(\mathbf{w}^+)}{2} +
```
This should be correct, `F` is the physical flux, i.e., the Euler operation, whereas `\hat{F}` the numerical flux on the left hand side.

bangerth replied:
Oh, good point. I missed that. It's hard to read latex source code :-)

kronbichler commented on examples/step-67/doc/intro.dox:
```diff
@@ -0,0 +1,680 @@
+The polynomial expansion of the solution is finally inserted to the weak form
+and test functios are replaced by the basis functions. This gives a
+discrete in space, continuous in time system with a finite number of unknown
+coefficient values $w_i$, $i=1,\ldots,n_\text{dofs}$. Regarding the choice of
```
> But also, we generally count from zero...

I can change that.

bangerth commented:
What's the status here?

kronbichler commented:
I am close to the finish line with my figures. One thing I realized is that the custom timer output that I added (and that @bangerth commented about) should rather go into a variant of `TimerOutput`. I think of a way to print the minimal, average and maximum of the times in various sections among the MPI ranks of an otherwise unsynchronized `TimerOutput` (given that the output is now adding barriers when we add a communicator). That would take me a few hours to finalize - I plan with it in the beginning of next week as I have teaching/exam duties this week.

bangerth commented:
Sounds like a good plan. I won't have time this week anyway, but wanted to make sure we don't get into the mad tutorial rush just before the release :-)

Ping me if you want me to take another look!

DavidSCN commented on examples/step-67/doc/intro.dox:
```diff
@@ -2,13 +2,695 @@
+  =
+  \begin{pmatrix}
+  \rho \mathbf{u}\\
+  \rho \mathbf{u} \otimes \mathbf{u}\\
```
I guess the pressure term is missing here for the diagonal terms in the velocity.

kronbichler replied:
Yes indeed - I fixed it now.

kronbichler commented:
This program is now ready.

bangerth commented:
There's a merge conflict. Can you resolve this?

bangerth commented:
I'm working through this program right now, directly making changes in a copy of the branch and will then push these changes directly (look, I'm 46 and still able to learn :-) ). Here are a few things that I can't address myself because I don't know the answer:
* In the formula for the CFL limit, there is a constant `Cr`. Can you describe what it is?
* In the low-storage ERK definition, you use intermediate variables named `t_i`. Is the scheme written there correct? I imagine it is, but can't quite convince myself that it is with the various `a` and `b` coefficients. Would appreciate if you could double-check the recursion statement. Also, maybe replace `t_i` by something that doesn't look like the time?

I've just created a pull request to your repo (I can't seem to push directly): https://github.com/kronbichler/dealii/pull/3 Please merge and, if you want, make further edits after the fact.

I'll then continue with the program and the results section.

kronbichler commented:
> In the low-storage ERK definition, you use intermediate variables named t_i. Is the scheme written there correct? I imagine it is, but can't quite convince myself that it is with the various a and b coefficients. Would appreciate if you could double-check the recursion statement. Also, maybe replace t_i by something that doesn't look like the time?

Let me double-check. Using another name is an excellent suggestion and a typical "not seeing the forest for the trees" effect - I had to choose a name that does typically not appear in the RK formulas so choose `t` for temporary vector. Let me see through the rest of the code which letter is still free and makes sense in the context.

kronbichler commented:
Thanks for the PR - I will look into that once I've resolved the names.

bangerth commented:
I think pretty much anything that doesn't actually appear in that formula is fine -- T_i would have been ok with me. The problem was simply that the formulas make explicit reference to the current time.

kronbichler commented:
I have addressed the remaining issues and further fixed a few things on top of @bangerth's comments. Thanks for the valuable input!

bangerth commented:
Can I ask you to, for the moment, not rebase things? Because then I can just keep pulling from your branch.

kronbichler commented:
> Can I ask you to, for the moment, not rebase things?

Sure, sorry for the mess-up.

bangerth commented:
On 3/24/20 9:44 AM, Martin Kronbichler wrote:
> Sure, sorry for the mess-up.

No worries -- it's not like you were supposed to know :-)

bangerth commented:
I'll have more pull requests against your patch here (https://github.com/kronbichler/dealii/pull/4). Here are a couple of questions that I don't have answers to and would like if you could address (after merging my PR into yours):
* What's the mesh resolution or number of cells/dofs for the pictures you show for the flow-around-the-cylinder benchmark?
* In the output in `results.dox`, you show `VECTORIZATION_LEVEL`. Is this still what you want to show after your patches over the past couple of days?
* There is a reference to Gassner 2013 that contains a direct link to a DOI. Do you mind putting that into references.bib?


More as a note (no need to change anything): When you plot partitions of a parallel mesh, it's often useful to choose a more chaotic color map than one with a gradient. In visit, for example, "rasmol" is an example. That's because it arranges 256 (or so) colors in a way so that the visual difference from one to the next is as large is possible, and you'll get a picture in which the partitions are more clearly delineated from each other, rather than one where each partition has a color different from its neighbor by only a hue, as is the case in the example you show.

kronbichler commented:
> More as a note (no need to change anything): When you plot partitions of a parallel mesh, it's often useful to choose a more chaotic color map than one with a gradient. In visit, for example, "rasmol" is an example. That's because it arranges 256 (or so) colors in a way so that the visual difference from one to the next is as large is possible, and you'll get a picture in which the partitions are more clearly delineated from each other, rather than one where each partition has a color different from its neighbor by only a hue, as is the case in the example you show.

That's a good comment - indeed, I had experimented with such a scheme, but found the present one more appropriate for what I want to show, namely where the low ranks are located versus the high ranks, not primarily the difference between the different subdomains.

Thanks for the comments and fixes otherwise - I have answered the other questions and updated the code.

bangerth commented:
Here's a couple more things about the program itself (another PR to come -- you might want to look through the commits individually for more clarity):

* Any reason why `FINAL_TIME` is all caps? No other variable is that way.
* I've changed all places in `EulerOperator` that use `std::shared_ptr` to `std::unique_ptr`. I think that's conceptually better, but it requires us to take a bunch of input arguments of the setter functions as non-const non-references. I'll let you decide whether that's what you want -- it's in a separate commit and so easily revertable.
* The various `EulerOperator::apply_*` functions all take a `MatrixFree` object as first argument -- but they're member functions, so they could also get at this variable via `this->`. Am I missing something? Are they call-backs from somewhere? Confusingly, both the argument and the member variable are called `data` :-)
* The `.cc` file contains a few references to `\Omega_e` as a cell. I *think* that the rest of the program/dox uses `K`. Can you double check and make it uniform?
* It is uncommon to copy the primal solution in `DataPostprocessor` objects. Rather, in all other cases I can think of one attaches the solution once via `add_vector`, and then a second time with a postprocessor argument. In other words, the postprocessor only computes *derived* quantities. There is nothing wrong with your solution, but it's not particularly idiosyncratic. One could also have written multiple postprocessors (which, in that case, can be derived from `DataPostprocessorScalar` and `DataPostprocessorVector` and would require less boiler-plate).
* Can you explain somewhere the formula used for the Schlieren plot?
* For testcase 0, you call `GridGenerator::subdivided_hyper_rectangle()` with a vector of subdivisions that consists of only ones. Is this a mistake? You have a square, so you don't have to subdivide more often -- but we have a function `GridGenerator::hyper_rectangle` :-)
* Where you use a locale to get 1,000-separators, you use `std::locale("en_US.UTF-8")`. Is it necessary to use `en_US.UTF-8`? step-32 just uses `""`.
* `compute_error` returns a `Tensor<1,3>` -- but it's not actually a tensor. Any reason not to use `std::array<3,double>`?

Finally, step-69 has this nice movie at the end. I wonder whether we could have something like this also for this program (maybe as a follow-up patch)? step-67 just directly links to it, but we could as well embed a link to a youtube video.

bangerth commented:
Other than the comments above and the PR just posted against your branch, I have nothing else -- looks good, thanks very much for the effort!

kronbichler commented:
> The various EulerOperator::apply_* functions all take a MatrixFree object as first argument -- but they're member functions, so they could also get at this variable via this->. Am I missing something? Are they call-backs from somewhere?

Yes, they are call-backs from the `MatrixFree::loop()` that passes along the data container to attach to in case the functions are not member functions. I have chosen to uniformly use the class member, omit the name of the first callback argument, and added a short comment.

kronbichler commented:
I addressed all the comments.

A few answers to your questions:
> It is uncommon to copy the primal solution in DataPostprocessor objects.

This is because of my obsession of not computing things twice when they show up in a profiler like for the degree 5 in 3d (in this case the interpolation of the solution onto the points of the patches). But I completely agree with the pedagogic aspect here, so I switched to the usual setup.

> For testcase 0, you call `GridGenerator::subdivided_hyper_rectangle()` with a vector of subdivisions that consists of only ones. Is this a mistake?

This is due to some experiments I did during development, but they are of course no longer necessary, so of course we should switch to the `hyper_rectangle`.

> Where you use a locale to get 1,000-separators, you use `std::locale("en_US.UTF-8")`. Is it necessary to use `en_US.UTF-8`?

I did some development on a machine that sets the German locale `de_DE.UTF-8` - which made me set things explicitly. Not sure what we prefer - users getting the same output as the tutorial shows, or in their familiar data format. For reference, most of my machines have the `en_US` locale for the numeric types but some `de_DE` types for addresses and other stuff. I do not really like the German separators (I also don't fully like the US one - if I had to pick one, I would choose `de_CH` but that might be my bias of having grown up close to Switzerland and reading Swiss technical literature as well, even though the official format in my home region is either `de_AT` or `it_IT`, the former being almost entirely `de_DE`). In any case, I now use the empty `""` specifier consistent with other tutorials.

> compute_error returns a `Tensor<1,3>` -- but it's not actually a tensor. Any reason not to use `std::array<3,double>`?

I switched to the array for the return type and plain arrays `double errors_squared[3]` because the `std::array` does not have an overload in `Utilities::MPI::sum()`.

kronbichler commented:
Now that everything is complete, I guess we rebase and squash (which means that I probably will just have one big commit with everything, given that things are pretty mixed up now)?

kronbichler commented:
As to a movie: That would be a nice addition indeed, but I won't do it before I return to my office after Corona as I don't want to pull over all data home with my ISP for visualization at home and also do not want to set up remote visualization.

bangerth commented:
On 3/28/20 4:07 AM, Martin Kronbichler wrote:
> Now that everything is complete, I guess we rebase and squash (which means 
> that I probably will just have one big commit with everything, given that 
> things are pretty mixed up now)?

We've often had many commits for tutorials. It's fine by me whichever way you 
want!

kronbichler commented:
Then let's go with the current version, so I don't have to poke around. Let me just add a changelog, then everything is ready.

bangerth commented:
On 3/28/20 4:02 AM, Martin Kronbichler wrote:
>     Where you use a locale to get 1,000-separators, you use
>     |std::locale("en_US.UTF-8")|. Is it necessary to use |en_US.UTF-8|?
> 
> I did some development on a machine that sets the German locale |de_DE.UTF-8| 

You mean, when you do |std::locale("")| it sets the de.DE locale? That's 
bizarre. But maybe you're right and we should just make it explicit. But 
again, it's not an important point.


> German separators (I also don't fully like the US one - if I had to pick one, 
> I would choose |de_CH| but that might be my bias of having grown up close to 
> Switzerland and reading Swiss technical literature as well, even though the 
> official format in my home region is either |de_AT| or |it_IT|, the former 
> being almost entirely |de_DE|).

Now I'm curious: How do the Swiss do it?

There's of course also the Indian approach that uses separators for every 
power of 10^4 instead of 10^3 :-)

kronbichler commented:
> Now I'm curious: How do the Swiss do it?

It's the thousands separators I like (and some even more off-topic things):
```
Number of degrees of freedom: 147'456 ( = 4 [vars] x 1'024 [cells] x 36 [dofs/cell/var] )
```
The decimal separator is like the English `.`.

bangerth commented:
Ah, interesting! Some recent version of C++ also allows to use `'` in numbers to make it easier to read large numbers. I believe that they can actually be used at positions of your choice and simply have no meaning at all.

bangerth commented:
Yay! Good work!
